config/v1/types_oauth.go

```diff
@@ -1,37 +1,420 @@
+
+const (
+	// GrantHandlerAuto auto-approves client authorization grant requests
+	GrantHandlerAuto GrantHandlerType = "auto"
+	// GrantHandlerPrompt prompts the user to approve new client authorization grant requests
+	GrantHandlerPrompt GrantHandlerType = "prompt"
+	// GrantHandlerDeny auto-denies client authorization grant requests
+	GrantHandlerDeny GrantHandlerType = "deny"
+)
+
+// GrantConfig holds the necessary configuration options for grant handlers
+type GrantConfig struct {
+	// Method: allow, deny, prompt
+	Method GrantHandlerType
+
+	// ServiceAccountMethod is used for determining client authorization for service account oauth client.
+	// It must be either: deny, prompt
+	ServiceAccountMethod GrantHandlerType
+}
+
+var ValidGrantHandlerTypes = sets.NewString(string(GrantHandlerAuto), string(GrantHandlerPrompt), string(GrantHandlerDeny))
+var ValidServiceAccountGrantHandlerTypes = sets.NewString(string(GrantHandlerPrompt), string(GrantHandlerDeny))
+
+// TokenConfig holds the necessary configuration options for authorization and access tokens
+type TokenConfig struct {
+	// AuthorizeTokenMaxAgeSeconds defines the maximum age of authorize tokens
+	AuthorizeTokenMaxAgeSeconds int32 `json:"authorizeTokenMaxAgeSeconds"`
+	// AccessTokenMaxAgeSeconds defines the maximum age of access tokens
+	AccessTokenMaxAgeSeconds int32 `json:"accessTokenMaxAgeSeconds"`
+	// AccessTokenInactivityTimeoutSeconds defined the default token
+	// inactivity timeout for tokens granted by any client.
+	// Setting it to nil means the feature is completely disabled (default)
+	// The default setting can be overriden on OAuthClient basis.
+	// The value represents the maximum amount of time that can occur between
+	// consecutive uses of the token. Tokens become invalid if they are not
+	// used within this temporal window. The user will need to acquire a new
+	// token to regain access once a token times out.
+	// Valid values are:
+	// - 0: Tokens never time out
+	// - X: Tokens time out if there is no activity for X seconds
+	// The current minimum allowed value for X is 300 (5 minutes)
+	AccessTokenInactivityTimeoutSeconds *int32 `json:"accessTokenInactivityTimeoutSeconds,omitempty"`
+}
+
+// IdentityProvider provides identities for users authenticating using credentials
+type IdentityProvider struct {
+	// Name is used to qualify the identities returned by this provider
+	Name string `json:"name"`
+	// UseAsChallenger indicates whether to issue WWW-Authenticate challenges for this provider
+	UseAsChallenger bool `json:"challenge"`
+	// UseAsLogin indicates whether to use this identity provider for unauthenticated browsers to login against
+	UseAsLogin bool `json:"login"`
+	// MappingMethod determines how identities from this provider are mapped to users
+	MappingMethod string `json:"mappingMethod"`
+	// Provider contains the information about how to set up a specific identity provider
+	Provider runtime.RawExtension `json:"provider"`
```

Comment:
@smarterclayton it's a significant change (and this API is mostly lift and shift so far), but I think we've had better luck with `VolumeSource` style apis with a discriminator (`type`) and every choice enumerated with pointers.  That lets skewed clients communicate (discriminator), makes help helpful (what can I put here), and makes decoding easier.  Agree?

Reply:
Ultimately this is an API that we completely control.  Is there ever a case where we are not capable of enumerating this list fully?  If so, that might be the `externalIdentity` type that takes a config map or reference to a CR

Reply:
RawExtension is not good on human facing doc, and this will be 100% human facing (as in, an admin wants to run `oc explain oauth` and see what it means.

Reply:
Enumerated each type. Please take a look, I'm not sold on the naming I have for the fields if you have any suggestions on how to make it more clear.
https://github.com/openshift/api/pull/129/files#diff-338a1fe290fa58fdc89fa17c00c1e347R124